bin/wasm-node/rust/src/platform.rs

```diff
@@ -106,183 +104,463 @@ impl smoldot_light_base::Platform for Platform {
-        if !connection.messages_queue.is_empty() || connection.closed_message.is_some() {
-            return future::ready(()).boxed();
+    fn wait_more_data(
+        StreamWrapper(stream_id, read_buffer): &mut Self::Stream,
+    ) -> Self::StreamDataFuture {
+        if read_buffer.buffer_first_offset < read_buffer.buffer.len() {
+            return async move {}.boxed();
         }
 
-        let listener = unsafe {
-            Pin::get_unchecked_mut(connection.as_mut())
-                .something_happened
-                .listen()
+        let something_happened = {
+            let mut lock = STATE.try_lock().unwrap();
+            let stream = lock.streams.get_mut(stream_id).unwrap();
+
+            if !stream.messages_queue.is_empty() || stream.closed {
+                return future::ready(()).boxed();
+            }
+
+            stream.something_happened.listen()
         };
 
-        listener.boxed()
+        something_happened.boxed()
     }
 
-    fn read_buffer(connection: &mut Self::Stream) -> Option<&[u8]> {
-        if let Some(buffer) = connection.messages_queue.front() {
-            debug_assert!(!buffer.is_empty());
-            debug_assert!(connection.messages_queue_first_offset < buffer.len());
-            Some(&buffer[connection.messages_queue_first_offset..])
-        } else if connection.closed_message.is_some() {
-            None
+    fn read_buffer(StreamWrapper(stream_id, read_buffer): &mut Self::Stream) -> Option<&[u8]> {
+        let mut lock = STATE.try_lock().unwrap();
+        let stream = lock.streams.get_mut(stream_id).unwrap();
+
+        if stream.closed {
+            return None;
+        }
+
+        if read_buffer.buffer_first_offset < read_buffer.buffer.len() {
+            return Some(&read_buffer.buffer[read_buffer.buffer_first_offset..]);
+        }
+
+        // Move the next buffer from `STATE` into `read_buffer`.
+        if let Some(msg) = stream.messages_queue.pop_front() {
+            read_buffer.buffer = msg;
+            read_buffer.buffer_first_offset = 0;
+            Some(&read_buffer.buffer[..])
         } else {
             Some(&[])
         }
     }
 
-    fn advance_read_cursor(connection: &mut Self::Stream, bytes: usize) {
-        let this = unsafe { Pin::get_unchecked_mut(connection.as_mut()) };
+    fn advance_read_cursor(
+        StreamWrapper(stream_id, read_buffer): &mut Self::Stream,
+        mut bytes: usize,
+    ) {
+        loop {
+            // Advance `read_buffer`.
+            {
+                let read_buffer_advance = cmp::min(
+                    read_buffer.buffer.len() - read_buffer.buffer_first_offset,
+                    bytes,
+                );
+
+                read_buffer.buffer_first_offset += read_buffer_advance;
+                bytes -= read_buffer_advance;
+            }
 
-        this.messages_queue_first_offset += bytes;
+            // Avoid the whole locking process if `bytes` is 0.
+            if bytes == 0 {
```

Comment:
so we only exit when we have exact number of `bytes` (i.e. `read_exact`)? what if stream is closed?

Reply:
This function is to discard the first N bytes from the read buffer. If you pass a N that is larger than the number of bytes in the read buffer you get a panic. There is no relation with the stream being closed or not.